src/sidebar/components/search-input.js

```diff
@@ -1,45 +1,92 @@
-  this.isLoading = () => store.isLoading();
+const useStore = require('../store/use-store');
 
-  this.inputClasses = function() {
-    return { 'is-expanded': self.alwaysExpanded || self.query };
-  };
+const Spinner = require('./spinner');
+
+/**
+ * An input field in the top bar for entering a query that filters annotations
+ * (in the sidebar) or searches annotations (in the stream/single annotation
+ * view).
+ *
+ * This component also renders a loading spinner to indicate when the client
+ * is fetching for data from the API or in a "loading" state for any other
+ * reason.
+ */
+function SearchInput({ alwaysExpanded, query, onSearch }) {
+  const isLoading = useStore(store => store.isLoading());
+  const input = useRef();
+
+  // The active filter query from the previous render.
+  const [prevQuery, setPrevQuery] = useState(query);
+
+  // The query that the user is currently typing, but may not yet have applied.
+  const [pendingQuery, setPendingQuery] = useState(query);
 
-  this.$onChanges = function(changes) {
-    if (changes.query) {
-      input.value = changes.query.currentValue;
-    }
+  const onSubmit = e => {
+    e.preventDefault();
+    // TODO - When the parent components are converted to React, the signature
+    // of the callback can be simplified to `onSearch(query)` rather than
+    // `onSearch({ $query: query })`.
+    onSearch({ $query: input.current.value });
   };
+
+  // When the active query changes outside of this component, update the input
+  // field to match. This happens when clearing the current filter for example.
+  if (query !== prevQuery) {
+    setPendingQuery(query);
+    setPrevQuery(query);
+  }
+
+  return (
+    <form className="search-input__form" name="searchForm" onSubmit={onSubmit}>
+      <input
+        className={classnames('search-input__input', {
+          'is-expanded': alwaysExpanded || query,
+        })}
+        type="text"
+        name="query"
+        placeholder={(isLoading && 'Loading…') || 'Search…'}
+        disabled={isLoading}
+        ref={input}
+        value={pendingQuery}
+        onInput={e => setPendingQuery(e.target.value)}
+      />
+      {!isLoading && (
+        <button
+          type="button"
+          className="search-input__icon top-bar__btn"
+          onClick={() => input.current.focus()}
+        >
+          <i className="h-icon-search"></i>
+        </button>
+      )}
+      {isLoading && <Spinner className="top-bar__btn" title="Loading…" />}
+    </form>
+  );
 }
 
-module.exports = {
-  controller: SearchInputController,
-  controllerAs: 'vm',
-  bindings: {
-    // Specifies whether the search input field should always be expanded,
-    // regardless of whether the it is focused or has an active query.
-    //
-    // If false, it is only expanded when focused or when 'query' is non-empty
-    alwaysExpanded: '<',
-    query: '<',
-    onSearch: '&',
-  },
-  template: require('../templates/search-input.html'),
+SearchInput.propTypes = {
+  /**
```

Comment:
What application state causes `alwaysExpanded` to be a particular value?

Reply:
The stream and single-annotation page always expand this field. Grep for `always-expanded`.